src/com/amazon/ion/impl/lite/IonStructLite.java

```diff
@@ -314,20 +309,30 @@ private IonStruct doClone(boolean keep, String... fieldNames)
             }
 
             clone = getSystem().newEmptyStruct();
+            clone._children = new IonValueLite[_children.length];
+            clone._child_count = 0;
             for (IonValue value : this)
             {
-                SymbolToken fieldNameSymbol = value.getFieldNameSymbol();
-                String fieldName = fieldNameSymbol.getText();
-                if (fields.contains(fieldName) == keep)
+                if (fields.contains(((IonValueLite) value)._fieldName) == keep)
                 {
+                    IonValueLite clonedChild = (IonValueLite) value.clone();
```

Comment:
We can avoid the cast for accessing `_fieldName` by looping directly over the struct's `_children` array rather than implicitly using the iterator. The cast of the return from `clone()` is needed not only to access internals but also so that it can be assigned into the clone's `_children` array of `IonValueLite`.

Avoiding the implicit iterator creation also has the nice side effect of further improving performance.

#### IonStruct.cloneAndRemove, removing the largest child value from log.ion, leaving only scalar children

| dataset                | baseline (us/op) | implicit iterator (us/op)       | direct loop (us/op)
| -----------------| ---------------|-----------------|-----------------|
| log.10n                 | 209791            | 58310 (-72%) | 50054 (-76%) |

#### IonStruct.cloneAndRetain, retaining only the largest child value from log.ion, removing all scalar children

| dataset                | baseline (us/op) | implicit iterator (us/op)       | direct loop (us/op)
| -----------------| ---------------|-----------------|-----------------|
| log.10n                 | 299065            | 201514 (-33%) | 198688 (-34%) |

#### IonStruct.cloneAndRemove, removing a large blob value from events.10n and retaining annotated fields

`events.10n` is a 7.5MB binary Ion file containing about 2000 structs with annotated field names. The bulk of each struct is a large blob.

| dataset                | baseline (us/op) | implicit iterator (us/op)       | direct loop (us/op)
| -----------------| ---------------|-----------------|-----------------|
| events.10n                 | 978            | 672 (-31%) | 518 (-47%) |

#### IonStruct.cloneAndRetain, retaining only two annotated fields

| dataset                | baseline (us/op) | implicit iterator (us/op)       | direct loop (us/op)
| -----------------| ---------------|-----------------|-----------------|
| events.10n                 | 968            | 662 (-32%) | 549 (-43%) |